ImGui: Include a default font file and add font size option

TryTwo opened the pull request:
/edit There are too many issues with allowing users to select any font.  Many look bad in imgui and non-monospaced fonts will not work with the auto-adjust window size to font size. So, I elected to include bitstream vera as a default font to use.

Example, I also made the graph resizable.
<details><summary>Example</summary>
<p>

![imgui](https://github.com/user-attachments/assets/6e603bc8-d7cb-495c-8065-7c57addf17f7)
</p>
</details> 

ImGui has needed better fonts since it was added. I recently figured out how to change the font and size, but noticed various elements needed to be adjusted if the font was going to change. I've been playing with it, but I don't want to spend too much time polishing it up unless there is interest in this approach.  If someone else wants to work on it, I hope this will provide a bit of a guideline.

To test this you need to place a .ttf font file in the Dolphin  User/Themes folder and name it "imgui_font.ttf". I think managing fonts is one reason why this has been so slow to be tackled. There doesn't appear to be a way to directly interface with our Qt font, and we still have to support non-Qt backends (right?). The simple solution is to include a font file with Dolphin that it can automatically use, but I don't know if anyone is comfortable with that, or how it'd be managed properly.  My alternative here is to just have the user copy a font file over.

I hacked the option into the Settings -> interface panel for now. I think we could allow any font size, but I tried to keep it simple for testing.

Also note we can skip this font size change for any ImGui element that already looks fine, or set multiple font sizes. If anyone wants to test various ImGui elements that I can't (netplay, achievement), I could tweak those to support various sizes too.

iwubcode commented:
Curious what exactly you are trying to solve?

>I recently figured out how to change the font and size, but noticed various elements needed to be adjusted if the font was going to change. I've been playing with it, but I don't want to spend too much time polishing it up unless there is interest in this approach

Personally, I don't think we should, while imgui is generally slow in releasing big features...when they release, they are generally really nice and have a lot of user testing.  The next release is said to have dynamic font support.  No one can say when it will _actually release_ but we've held out this long...

TryTwo commented:
1.  Have bigger UI text, like a bigger FPS display.
2.  Make things like this look way better and more readable:
![image](https://github.com/user-attachments/assets/18389dfb-a6c7-445c-85c2-cdff3b2e6d92)

iwubcode commented:
>/edit Oh I see what it is, it's nice but not much different from what can already be done. We can already load multiple font sizes and switch between them. Not sure we would need dynamic fonts, except as an easy way to rescale font. It still requires you to load your own font for it to work.

Unless I'm misunderstanding, the new feature in imgui is exactly what you are attempting to do in this PR.  Scale a font dynamically?

> I recently figured out how to change the font and size, but noticed various elements needed to be adjusted if the font was going to change.

This is what concerns me the most.  My hope (maybe incorrect?) is that with the imgui update this will all be seamless.

---

I guess it depends on how much impact this change has.  I haven't looked at your code in any detail.

TryTwo commented:
The only messy thing I did was throw the option into the InterfacePane. Everything else about using fonts and sizes is straightforward with a few lines of code that won't be improved much by future updates. ImGui already automatically sizes most things to your font size if you don't force a size.

No matter what, we need to add a font that can be properly scaled,. I'm not trying to scale it dynamically, I'm just setting a font size for everything, and if that breaks something we can use a different font size, which sounds dynamic but really isn't. It's 3 lines of code.

We also have some things with hardcoded magic-number sizes, which I just changed to being auto-calculated or based on font size. I doubt they'd work differently later.

Choosing a font for every OS is the only pain, and that won't change unless we package a font with dolphin.

Pick font sizes:
AddFontFromTTF("font", size1)
AddFontFromTTF("font", size2)

Set default font for everything:
.FontDefault = font1

Change the font for a certain element:
PushFont(font2)
PopFont()

TryTwo commented:
Oh sorry, I see the point now.  Dynamic Fonts doesn't require you to re-run the texture/GPU code to get a new font size.

What I'm doing is loading multiple font sizes before they turning it into GPU textures. Then you just have multiple font sizes you can switch between whenever.  If there were to be finer control over the font size, like choosing any value, then you would either:
- Only allow it when the GPU isn't running
- Re-run the code to turn the fonts into a texture while the GPU is running. Not sure if there's any trouble with that.
- Wait for Dynamic Fonts.

I don't think any of that impacts this PR much. My font texture is just bigger because it has multiple sizes in it. The main points are doing minor UI tweaks to support the font sizes, and deciding on how to let users use a scalable font.

TryTwo commented:
Added font file/path selector.
Changed it to update the font between frames, so it can change fonts and sizes while running.
It only needs two font sizes now. An automatically set 14-point for small things like graph labels, and the user selected size. I could make the small font be the imgui default font, but I set it to be based on the custom font chosen.

That Dynamic Font Size imgui update will only skip the call to SetFont() when changing the font size while a game is running. It'll still be needed if changing the font itself.

mbc07 commented:
I'm not sure if the new settings this PR adds belongs to Config => Interface. All usages of ImGui on Dolphin currently rely on the the renderer window being visible/active, so Graphics => General seems like a more suitable location for ImGui-related settings?

TryTwo commented:
@mbc07  Any thoughts on making this not terrible?  Is there a world where putting the spinbox on the same line as the path makes sense? Should I just put a 50% spacer on  the right side of the size?
![image](https://github.com/user-attachments/assets/1e47fde6-115c-4daa-90dc-b7f341f2b207)

mbc07 commented:
I'm not sure in how to improve this one, but here are some food for thought...

> Is there a world where putting the spinbox on the same line as the path makes sense?

I don't think we have precedents for that. You could try changing the label to "Imgui Font Path/Size:" and then add the spinbox to the right of the "..." button, but I'm not sure if that would look better/worse than the current iteration of this PR.

> Should I just put a 50% spacer on the right side of the size?

Normally the widgets are aligned either to the right or to the left of the parent panel. With that said, if you want to add a spacer, it should likely go between the label and the spinbox, not to the right of the spinbox. You could also try changing that panel to a two column layout, like the Config => Wii tab or any of the graphics tabs. The spinbox would still end large, but a bit less. Perhaps another option would be replacing the spinbox with a slider (like the ones from the Audio tab)?

OatmealDome commented:
> The simple solution is to include a font file with Dolphin that it can automatically use, but I don't know if anyone is comfortable with that, or how it'd be managed properly.

I feel like we should strongly consider this if the default font is inadequate. [afaik, we already include a free substitute for the font found in the GC IPL](https://dolphin-emu.org/blog/2016/01/02/dolphin-progress-report-december-2015/#40-8488-always-load-system-fonts-from-ipl-dump-if-available-by-jhonn), so I don't see why we can't include one for ImGui.

TryTwo commented:
I'm not sure how to approach it, so someone else would need to take it on. I'm guessing we can't just tap into the font we include for the GC IPL you mentioned?

OatmealDome commented:
The IPL font is stored in a custom format in `Sys/GC/font_*.bin`, so I don't think it will be usable for this PR.

Can't we just ship a TTF file in the Sys directory (given that the font is properly licensed, of course)?

TryTwo commented:
Okay, maybe I was over thinking things.  

Two small (~50kb), monospaced, .ttf I can find are Cousine, which is an example font in the imgui project and Bitstream Vera, which is a base for various common fonts.

If we included it, what folder would it go in? Options: new folder, Resources, Themes, the base directory.
